File: aoc03.py
# Read input

with open("input03.txt") as f:
    inputs = f.readlines()

inp = [s.strip() for s in inputs]

l = len(inp)

m = len(inp[0])

# Part one

cnt = []

for i in range(m):
    c = 0
    for j in range(l):
        if inp[j][i] == '1':
            c+=1
    if c > l//2:
        cnt.append(1)
    else:
        cnt.append(0)

gamma = 0
epsilon = 0

for c in cnt:
    if c == 0:
        gamma *= 2
        epsilon = 2*epsilon + 1
    else:
        gamma = 2*gamma+1
        epsilon *= 2

print(gamma*epsilon)


# Part two

oxinds = set(range(l))

pos = 0

while len(oxinds) > 1:
    cnt = 0
    inds1 = set([])
    inds0 = set([])
    for i in oxinds:
        if inp[i][pos] == '1':
            cnt+=1
            inds1.add(i)
        else:
            inds0.add(i)
    if 2*cnt >= len(oxinds):
        oxinds = oxinds - inds0
    else:
        oxinds = oxinds - inds1
    pos += 1

co2inds = set(range(l))

pos = 0

while len(co2inds) > 1:
    cnt = 0
    inds1 = set([])
    inds0 = set([])
    for i in co2inds:
        if inp[i][pos] == '0':
            cnt+=1
            inds0.add(i)
        else:
            inds1.add(i)
    if 2*cnt <= len(co2inds):
        co2inds = co2inds - inds1
    else:
        co2inds = co2inds - inds0
    pos += 1

def bin2dec(ds):
    r = 0
    for d in ds:
        if d=='0':
            r*=2
        else:
            r = 2*r + 1
    return(r)

ox = list(oxinds)[0]
co2 = list(co2inds)[0]

print(bin2dec(inp[ox])*bin2dec(inp[co2]))

File: aoc03.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func bin2dec(ds []int) int {
	res := 0
	for _, d := range ds {
		if d == 0 {
			res *= 2
		} else {
			res = 2*res + 1
		}
	}
	return res
}

func main() {
	file, err := os.Open("input03.txt")
	if err != nil {
		fmt.Println(err)
	}
	defer file.Close()

	inp := make([]string, 0)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		inp = append(inp, scanner.Text())
	}

	l := len(inp)
	m := len(inp[0])

	sums := make([]int, 0)

	for j := 0; j < m; j++ {
		cnt := 0
		for i := 0; i < l; i++ {
			if inp[i][j] == '1' {
				cnt += 1
			}
		}
		sums = append(sums, cnt)
	}

	gamma := make([]int, 0)
	epsilon := make([]int, 0)

	for j := 0; j < m; j++ {
		if sums[j] > l/2 {
			gamma = append(gamma, 1)
			epsilon = append(epsilon, 0)
		} else {
			gamma = append(gamma, 0)
			epsilon = append(epsilon, 1)
		}
	}

	fmt.Println(bin2dec(gamma) * bin2dec(epsilon))

}
